### ISSUE TITLE
[Question] Use as part of sub-package

### ISSUE THREAD

**Issue by m-Skolnick** [dart-code-checker/dart-code-metrics#254]
I am trying to use `dart_code_metrics` in my analyzer package which I have as a sub-package to my main project.

Is this possible? 

VSCode does not seem to pick up the additional rules unless I put it in my `main_package/analysis_options.yaml`.

My project structure is as follows:

main_package/
- pubspec.yaml
- analysis_options.yaml
- packages / my_analysis_package

I then include the analysis options from my sub-package in `main_package/analysis_options.yaml` as follows:

```
include: package:my_analysis_package/analysis_options.yaml
```

The typical dart analyzer rules work as expected. This package just doesn't seem to be picked up.

**Comment by dkrutskikh** [comment]
@m-Skolnick please provide version of `dart_code_metrics`

**Comment by m-Skolnick** [comment]
> @m-Skolnick please provide version of `dart_code_metrics`

I have only tried with 2.4.1.

**Comment by incendial** [comment]
@dkrutskikh Am I correct, that we don't support `include` syntax? It seems like a decent feature and I think we should add it.

**Comment by incendial** [comment]
Nope, I was wrong, we even have tests for it.

@m-Skolnick could you please share a reproducible example?

**Comment by incendial** [comment]
Thank you, I will take a look.

**Comment by incendial** [comment]
Looks like `include` works fine, the problem is with the plugin discoverability:
```yaml
analyzer:
  plugins:
    - dart_code_metrics
```

If you add `dart_code_metrics` dependency to a sub package, then a root package analysis_options.yaml `include` will lead to a plugin manager trying to find `dart_code_metrics` for the root package.

In other words I see two quick solutions here:
1. put the config in `main_package/analysis_options.yaml` and have a root dependency as you suggested.
2. add `dart_code_metrics` dependency to the root package (or, if you don't  want whole project to be analysed, to every sub package), but keep all config in a singe sub package.

Both solutions looks not that great to me, but I don't think that we can solve it fast because it's about integration with the analyzer.

What do you think?

**Comment by m-Skolnick** [comment]
@incendial Thank you for looking into that! I want to be able to use the same analysis standards in multiple packages, so your solution #2 sounds like a good work-around for now! 

For anyone with the same issue, you just need to add the `dart_code_metrics` dependency wherever you use the analysis package. 

For example:

```
dev_dependencies:
  dart_code_metrics: ^2.4.1
  my_analysis_package:
    path: packages/my_analysis_package
```

Thanks for your help!

**Comment by incendial** [comment]
Great! Thank you for the example, it was very helpful!

**Comment by O-Hannonen** [comment]
I'm struggling with the exact same situation. I have my linting rules in a separate package. Only difference to @m-skolnick 's setup is that my linting rules are in a separate git repository, which the project itself is depending on. When trying the solution 1 described above, everything works. But that is not something I want to do, since I want to have consistent linting rules across all of my projects without copy pasting anything. Solution 2 described above would be a good workaround to solve this, except it does not work... 

When I have the project setup according to solution 2, my VSCode analyzer server does not recognize any plugins being active (have already tried all the steps in TROUBLESHOOTING.md). 

Just to make sure I understood  @incendial 's suggested solution (2. one) right: 
1. In `main_package/analysis_options.yaml`, there should be just the reference to the linter packages yaml `include: package:analysis_package/analysis_options.yaml`?
2. In `analysis_package/pubspec.yaml`, `dart_code_metrics` should be as a dependency?
3. In `main_package/pubspec.yaml`, `dart_code_metrics` should be as a dependency?
4. In `analysis_package/analysis_options.yaml`, `dart_code_metrics` should be defined as analyzer plugin, and in the same file all the `dart_code_metrics` configurations should be set?

**Comment by incendial** [comment]
Yeah, if I'm not mistaken, it should work. Here is an example of a reusable config with DCM https://github.com/MewsSystems/mews-flutter/blob/master/mews_pedantic/lib/analysis_options.yaml.

@ookami-kb hi! Maybe you can spot what's missing in the steps above?

**Comment by ookami-kb** [comment]
Step 2 is not really needed (unless you want to add some custom scripting on top of `dart_code_metrics` or use it in other way in the `analysis_package` itself).

But in `main_package/pubspec.yaml`, `analysis_package` itself should be added as dependency.

Also, the options file should be located in `analysis_package/lib/analysis_options.yaml` (but you still should reference it via `include: package:analysis_package/analysis_options.yaml` without `lib`).

Everything else looks correct.

**Comment by O-Hannonen** [comment]
Yes that is exactly what Im doing. Does it matter in what order the stuff is put into `analysis_options.yaml`? The order is the only thing I have different in my setup. I'll try to create a new project to test it, in case the issue is related to my current project setup.

**Comment by ookami-kb** [comment]
No, order shouldn't matter.

**Comment by O-Hannonen** [comment]
Just figured it out. For some reason `flutter pub get` was not fetching the newest version of the analysis package from github, even when running `flutter clean` before it. The package reference in `pubspec.yaml` was pointing to a specific branch on the repo. Making the reference point to a specific commit instead of a branch fixed the issue. I don't know if its related to my setup or just if its just the way flutter handles packages referenced from github repos, but at least I got it working now. Thanks for the great linting rules and quick responses!
